### ISSUE TITLE
broken tinygrad responses

### ISSUE THREAD

**Issue by roryclear** [exo-explore/exo#500]
I'm getting bad responses from tinygrad (except for the first). Running on an M2 Mac Mini, I've hardcoded the inference_engine_name value to use tinygrad instead of MLX. Seeing the same thing happen for 3B too, I haven't tried other models.
Will update if I find more info.

MLX output:
![image](https://github.com/user-attachments/assets/c47e7c57-073a-43b4-bc56-5c095979a4ac)

Tinygrad output:
![image](https://github.com/user-attachments/assets/d25142e7-4b2d-48ef-ba1b-4ac1cca7e981)

**Comment by roryclear** [comment]
in the tinygrad screenshot it hasn't answered what I've asked in the second prompt at all. Try having a conversation with 1B using MLX and then tinygrad, I'm just getting nonsense and/or irrelevant responses after the first, it's like it's not receiving the prompts correctly or something.

Didn't realize the models were different, if this is just down to the models being different then obv not an issue, does feel like more than a slight difference though

**Comment by AlexCheema** [comment]
> in the tinygrad screenshot it hasn't answered what I've asked in the second prompt at all. Try having a conversation with 1B using MLX and then tinygrad, I'm just getting nonsense and/or irrelevant responses after the first, it's like it's not receiving the prompts correctly or something.
> 
> Didn't realize the models were different, if this is just down to the models being different then obv not an issue, does feel like more than a slight difference though

The example you gave has context of the previous part of the conversation.
Can you give an example where it doesn't have context of the previous part?

**Comment by roryclear** [comment]
ah this might explain more
MLX:
![image](https://github.com/user-attachments/assets/b568d448-1c54-4263-b531-b3f0b0c3e06f)
Tinygrad:
![image](https://github.com/user-attachments/assets/aa3ca850-6bc5-4e29-a04c-e13521922897)

**Comment by blindcrone** [comment]
@AlexCheema Yea, this looks like a context bug to me, and makes an argument for spending some time reconciling the different caching methods between these implementations, and fully utilizing the caching for inference over a chat session rather than stacking prompts within the API. This will also fix some bugs that happen during a long session due to passing too large a context into the inference model

**Comment by AlexCheema** [comment]
> @AlexCheema Yea, this looks like a context bug to me, and makes an argument for spending some time reconciling the different caching methods between these implementations, and fully utilizing the caching for inference over a chat session rather than stacking prompts within the API. This will also fix some bugs that happen during a long session due to passing too large a context into the inference model
> 
> This kind of change will dovetail well with some other inference engine compatibility generalizations, so it seems like a good thing to do

Agree, that seems like a good strategy to fix this and probably a bunch of other unknown bugs.

**Comment by roryclear** [comment]
I found this doesn't happen with JIT off, I'm going to try narrow it down now. I think it's likely a bug in tinygrad, I've found a similar bug before with it

**Comment by Evanev7** [comment]
Stale for 1.0
